fix: feignclient 사용하여 업체 서비스의 각 비즈니스 로직을 권한 별 처리하도록 수정

yezyaa opened the pull request:
## ✏️ 작업한 내용을 간략히 설명해 주세요!
- 업체 서비스 예외 처리를 위해 FeignClient를 통해서 사용자 및 허브 서비스 API 호출
- 업체 서비스 각 비즈니스 로직을 권한 별(마스터 관리자, 허브 관리자, 허브 업체)로 처리 되게 수정

## 🛠️ 변경을 위해 진행한 작업 내용에는 어떤 것이 있나요?

- 📌 FeignClient를 통해 외부 API 호출 후 예외 처리
- 📌@PageableDefault 사용하여 페이징 파라미터 단순화
- 📌 파사드 디자인 패턴 적용하여 비즈니스 로직 권한 별 처리
- 📌 업체 조회 시, 마스터 관리자는 삭제된 업체도 조회 가능하도록 처리
- 📌 허브 관리자와 허브 업체는 자신이 소속된 업체에 대해서만 전체 조회 및 수정 가능
- 📌 @SuperBuilder를 사용하여 상속된 DTO 필드 처리
- 📌 업체 수정 후 다시 조회하는 순서가 괜찮다고 생각하여 리턴을 ResponseDto에서 Void로 변경
- 📌 업체의 hubId를 수정 시, 관리자만 수정 되도록 처리
- 📌JPA DirtyChecking을 활용하여 Entity가 EntityManager에 의해 관리되도록 수정

## 📝 변경 사항을 확인하기 위해 테스트한 방법을 설명해 주세요.
- company-app-user.http, company-app.http에서 각 API의 구동 테스트
- API 호출 후 DB에서 결과 확인

## 💬 리뷰 요구사항

- 예외 처리로 코드가 너무 길어지는 것 같아 중괄호를 많이 생략했는데 괜찮은 방법인지 궁금합니다.
  - 중괄호 추가함

## 📚 참고 자료

> 구현에 참고한 자료가 있다면 공유해주세요!

---

yezyaa committed:
config: FeignClient 활성화를 위한 설정 추가

yezyaa committed:
feat: 허브 서비스 FeignClient 연결 및 에러코드 추가

- 업체 생성 시, 허브가 존재하지 않을 때 에러 발생

yezyaa committed:
feat: 사용자 서비스 FeignClient 연결 및 에러코드 추가

- 업체 생성 시, 사용자가 존재하지 않을 때 에러 발생

yezyaa committed:
feat: 업체 등록 시, FeignClient를 통해 허브Id와 사용자Id 존재 여부 확인하여 예외 처리 추가

yezyaa committed:
fix: 사용자 권한 예외 처리 로직 수정

yezyaa committed:
fix: 허브 관리자가 소속된 허브의 업체만 삭제할 수 있도록 수정

- 마스터 관리자는 모든 업체를 삭제할 수 있도록 처리
- 허브 관리자는 소속된 허브의 업체만 삭제 가능하도록 'hubId' 비교 추가
- 삭제된 업체와 삭제되지 않은 업체 조회를 위한 쿼리 메서드 추가

yezyaa committed:
fix: 업체 전체 조회 및 검색 기능을 권한 별로 처리

- 마스터 관리자 : 모든 업체 조회 가능 (삭제된 업체 포함)
- 허브 관리자, 허브 업체 : 삭제되지 않은 업체 + 소속 허브 업체만 조회
- 관련된 쿼리 메서드 추가
- Lombok @SuperBuilder 적용을 통해 상속된 DTO 필드 처리 문제 해결
- 마스터 관리자용 CompanyDetailResponse DTO 추가 : isDeleted, updatedAt, updatedBy 추가
- Lombok @SuperBuilder 적용을 통해 상속된 DTO 필드 처리 문제 해결
- api test 추가

yezyaa committed:
fix: 업체 단일 조회 기능을 권한 별로 처리

- 마스터 관리자 : 모든 업체 조회 가능 (삭제된 업체 포함)
- 허브 관리자, 허브 업체 : 삭제되지 않은 업체 + 소속 허브 업체만 조회
- 관련된 쿼리 메서드 추가
- api test 추가

yezyaa committed:
fix: 업체 수정 기능을 권한 별로 처리

- 리턴 값을 Void로 수정 : 수정 후 다시 조회하는 방향이 좋다고 생각
- 마스터 관리자 : 모든 업체 수정 가능 (삭제된 업체 포함)
- 허브 관리자, 허브 업체 : 삭제되지 않은 업체 + 소속 허브 업체만 수정 가능
- hubId 수정은 관리자만 가능하도록 예외 처리
- api test 추가

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyService.java:
```diff
@@ -21,11 +31,16 @@
         validateAdminAuthority(role);
 
+        checkHubExists(registerCompanyRequest.getHubId());
+        checkUserExists(registerCompanyRequest.getOwnerId());
```
👍 👍

re-mora commented on service/company-app/src/main/java/io/fulflix/company/api/dto/CompanyResponse.java:
```diff
@@ -3,12 +3,12 @@
 
 @Getter
-@Builder
+@SuperBuilder
```
상속된 필드의 빌더 사용을 위해 @SupberBuilder를 사용하셨군요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/api/CompanyController.java:
```diff
@@ -48,31 +49,57 @@ public ResponseEntity<Page<CompanyResponse>> getAllCompanies(
     ) {
         CompanyResponse updatedCompany = companyService.updateCompany(id, updateCompanyRequest, currentUser, role);
-        return ResponseEntity.ok(updatedCompany);
+        return ResponseEntity.noContent().build(); // 204 No Content 응답
```
수정에 대한 응답으로 204를 반환하신 이유가 궁금합니다! 👀

yezyaa replied:
수정 후 수정된 내용을 볼 때는 다시 조회를 해야한다고 생각해서 리턴을 받지 않았습니다. 수정에 대한 응답을 반환받는게 더 좋을까요?

re-mora replied:
어떤 의도로 구성하신건지 궁금했어요! 이해 했습니다.! 👍 👍 
반환값이 없고 해당 요청을 성공적으로 처리했으면 204 No Content가 적절한 것 같아요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/CompanyApp.java:
```diff
@@ -17,6 +18,7 @@
         )
 })
 @EnableAsync
+@EnableFeignClients(basePackages = "io.fulflix.infra.client")
```
해당 설정이 변경되면 하드 코딩으로 적용된 설정을 찾아서 변경해줘야 하는 번거로움이 있을 것 같아요. 이를 위해 app-core의 FeignClientConfig에서 아래와 같이 상수를 제공하고 있으니 적용해보시는 것도 좋을 것 같아요!

```java
@Configuration
public class FeignClientConfig {

    private static final String FEIGN_CLIENT_PACKAGE = "infra.client";
    public static final String FEIGN_CLIENT_BASE_PACKAGE =
        BASE_PACKAGE + "." + FEIGN_CLIENT_PACKAGE; // io.fulflix.infra.client

    @Bean
    public RequestHeaderInterceptor requestHeaderInterceptor() {
        return new RequestHeaderInterceptor();
    }

}

```

re-mora commented on service/company-app/src/main/java/io/fulflix/company/api/CompanyController.java:
```diff
@@ -48,31 +49,57 @@ public ResponseEntity<Page<CompanyResponse>> getAllCompanies(
             @CurrentUserRole Role role
     ) {
-        Page<CompanyResponse> companies = companyService.getAllCompanies(query, page, size, sortBy, sortDirection, currentUser, role);
+        Page<CompanyDetailResponse> companies = companyService.getAllCompaniesForAdmin(query, page, size, sortBy, sortDirection, currentUser, role);
```
page, size, sortBy, sortDirection 모두@PageableDefault 어노테이션에서 제공하고 있으니, 적용해보시는 것 도 좋을 것 같아요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/api/CompanyController.java:
```diff
@@ -48,31 +49,57 @@ public ResponseEntity<Page<CompanyResponse>> getAllCompanies(
+            @RequestParam(defaultValue = "10") int size, // 페이지 크기
+            @RequestParam(defaultValue = "createdAt") String sortBy, // 정렬 기준
+            @RequestParam(defaultValue = "desc") String sortDirection, // 정렬 방향
+            @RequestParam(defaultValue = "0") int page, // 페이지 번호
```
이 부분도 page, size, sortBy, sortDirection 모두 제공하고 있는 @PageableDefault 어노테이션을 적용해보시면 좋을 것 같아요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyService.java:
```diff
@@ -36,46 +51,93 @@ public void registerCompany(RegisterCompanyRequest registerCompanyRequest, Long
+    public Page<CompanyDetailResponse> getAllCompaniesForAdmin(String query, int page, int size, String sortBy, String sortDirection, Long currentUser, Role role) {
+        validateMasterAdminAuthority(role);
+
+        if (size != 10 && size != 30 && size != 50) size = 10;
```
페이지 사이즈가 10/30/50으로 고정되어야 하는 요구사항을 만족하기 위해 모든 서비스가 해당 분기문을 고려하지 않고도,구성할 수 있는 방법을 고민해보셨으면 좋겠어요!

Spring에는 비지니스 로직 외의 공통 관심사를 처리할 수 있는 AOP를 제공하고 있어요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyService.java:
```diff
@@ -36,46 +51,93 @@ public void registerCompany(RegisterCompanyRequest registerCompanyRequest, Long
+        // 허브 관리자, 허브 업체 : 삭제되지 않은 업체 + 소속 허브의 업체만 조회
+        Page<Company> companies;
+
+        if (isHubAdmin(role)) {
```
업체 전체 조회 시, 
1. 관리자일 때 검색어가 있는 경우와 없는 경우
2. 허브 관리자일 때 검색어가 있는 경우와 없는 경우

를 처리하기 위한 분기문이 복잡해 수정하면서 실수가 발생할 여지가 있고, 지속적으로 코드를 관리가 어려울 수 있을 것 같아요

이를 해결하기 위해 전략 패턴을 적용해본다면
1. 관리자일 때의 요구사항을 만족하는 업체 목록을 조회 부
2. 허브 관리자일 때의 요구사항을 만족하는 업체 목록 조회 부

를 interface를 기준으로 각각 구현하여, 요청 권한에 따라 적절한 구현체를 주입하여 해결해 볼 수 있을 것 같아요!

---
추가적으로 중괄호 없이 if-else 분기문을 표현하는 경우 `Dangling else problem`가 발생할 수 있음을 살펴보시면 좋을 것 같아요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyService.java:
```diff
@@ -36,46 +51,93 @@ public void registerCompany(RegisterCompanyRequest registerCompanyRequest, Long
-        if (updateCompanyRequest.getHubId() != null && !updateCompanyRequest.getHubId().equals(company.getHubId()))
+        Company company;
+
+        if (isMasterAdmin(role))
```
이 부분도 마찬 가지로 전략 패턴을 이용해 요청에 적절한 전략을 주입하는 방향으로 코드 가독성과 유지보수성을 향상 시킬 수 있을 것 같아요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyService.java:
```diff
@@ -90,50 +152,106 @@ public CompanyResponse updateCompany(Long id, UpdateCompanyRequest updateCompany
 
-        Company company = findCompanyById(id);
-        company.delete(); // isDeleted = true
+        if (isMasterAdmin(role)) {
```
이 부분에도 전략 패턴을 적용해서 권한별 적절한 요구사항을 만족하는 구현체를 주입함으로써 코드 복잡성을 떨어트리고 유지보수성을 향상 시킬 수 있을 것 같아요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyService.java:
```diff
@@ -90,50 +152,106 @@ public CompanyResponse updateCompany(Long id, UpdateCompanyRequest updateCompany
+    }
+
+    // 마스터 관리자 권한 확인 예외처리
+    private void validateMasterAdminAuthority(Role role) {
```
꼼꼼한 예외 처리 👍 👍

re-mora commented on service/company-app/src/main/java/io/fulflix/infra/client/hub/HubClient.java:
```diff
@@ -0,0 +1,21 @@
+import static io.fulflix.infra.client.hub.HubClient.HUB_APP_CLIENT;
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+
+@FeignClient(HUB_APP_CLIENT)
```
호출하는 hub-app에 `현재 인증된 사용자의 번호와 권한`이 필요한 경우 FeignClient를 통해 어떻게 hub-app에 전달해 줄 수 있을까요?

re-mora commented on service/company-app/src/main/java/io/fulflix/infra/client/user/UserClient.java:
```diff
@@ -0,0 +1,21 @@
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+
+@FeignClient(USER_APP_CLIENT)
+public interface UserClient extends UserService {
```
이 부분도 호출하는 user-app에 현재 인증된 사용자의 번호와 권한이 필요한 경우 FeignClient를 통해 어떻게 user-app에 전달해 줄 수 있을지 고민이 필요할 것 같아요!

re-mora reviewed (changes requested):
전반적으로 조금 복잡할 수 있는 요구사항들을 적용하기 위해 유효성 검사와 예외 처리가 꼼꼼하게 적용되어 있고, 작성된 코드가 깔끔해서 읽는데 큰 어려움이 없었어요!

생각보다 손이 많이 가는 작업이셨을텐데 고생 많으셨습니다! 👍 

다만 여러 요구사항을 적용하는 과정에서 고민해보면 좋을만한 내용을 남겼습니다. 확인 부탁드려요!

re-mora committed:
refactor: add company retrieve facade and strategy

yezyaa committed:
refactor: FeignClient 설정을 상수 기반 공통 설정으로 리팩토링

yezyaa committed:
refactor: 모든 조건문에 중괄호 추가하여 가독성 및 안정성 향상

yezyaa committed:
refactor: @PageableDefault 사용으로 페이징 파라미터 단순화

- HTTP 클라이언트 테스트로 리팩토링 결과 확인
  - sort에 createdAt만 들어갈 경우, 기본값을 내림차순으로 설정했으나 Spring Data JPA의 기본 동작에 따라 오름차순으로 정렬됨

yezyaa committed:
refactor: 업체 전체 조회 및 검색 API에서 파사드 패턴 적용

- CompanyFacade 클래스를 구현하여 권한에 맞는 조회 전략 적용
- CompanyRetrieveStrategy 인터페이스 정의
- HubAdminRetrieveStrategy, HubCompanyRetrieveStrategy 전략 클래스 구현
  - 허브 관리자 및 허브 업체에 따른 조회 전략 적용

yezyaa committed:
refactor: 업체 단일 조회 API에서 파사드 패턴 적용

- CompanyFacade 클래스를 구현하여 권한에 맞는 조회 전략 적용
- CompanyRetrieveByIdStrategy 인터페이스 정의
- HubAdminRetrieveByIdStrategy, HubCompanyRetrieveByIdStrategy 전략 클래스 구현
  - 허브 관리자 및 허브 업체에 따른 조회 전략 적용

yezyaa committed:
refactor: 업체 수정 API에서 파사드 패턴 적용

- CompanyFacade 클래스를 구현하여 권한에 맞는 조회 전략 적용
- CompanyUpdateStrategy 인터페이스 정의
- MasterAdminUpdateStrategy, HubAdminUpdateStrategy, HubCompanyUpdateStrategy 전략 클래스 구현
  - 마스터 관리자, 허브 관리자, 허브 업체에 따른 조회 전략 적용
  - 파사드 패턴을 적용하여 마스터 관리자의 전략 클래스에만 hubId 수정 가능하게 함
- 업체 수정에서 쓰일 공통 예외 처리 클래스 추가

yezyaa committed:
refactor: 업체 삭제 API에서 파사드 패턴 적용

- CompanyFacade 클래스를 구현하여 권한에 맞는 조회 전략 적용
- CompanyDeleteStrategy 인터페이스 정의
- MasterAdminUpdateStrategy, HubAdminUpdateStrategy 전략 클래스 구현
  - 마스터 관리자, 허브 관리자에 따른 조회 전략 적용
- 기존 companyService에 있던 비즈니스 로직 삭제

yezyaa commented:
파사드 패턴 사용해서 각 API마다 권한에 맞는 조회,수정,삭제 전략 적용했습니다.

yezyaa committed:
chore: 사용하지 않는 import 정리

yezyaa committed:
Merge branch 'develop' into feature/connect-feignclient-to-company

re-mora commented on service/company-app/src/main/java/io/fulflix/company/CompanyApp.java:
```diff
@@ -18,7 +19,7 @@
 })
 @EnableAsync
-@EnableFeignClients(basePackages = "io.fulflix.infra.client")
+@EnableFeignClients(basePackages = FEIGN_CLIENT_BASE_PACKAGE)
```
👍

re-mora commented on service/company-app/src/main/java/io/fulflix/company/api/CompanyController.java:
```diff
@@ -42,29 +45,23 @@ public ResponseEntity<Void> registerCompany(
-            @RequestParam(defaultValue = "createdAt") String sortBy, // 정렬 기준
-            @RequestParam(defaultValue = "desc") String sortDirection, // 정렬 방향
-            @RequestParam(defaultValue = "0") int page, // 페이지 번호
+            @PageableDefault(sort = "createdAt", direction = Sort.Direction.DESC) Pageable pageable, // 페이징 및 정렬
```
👍

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyRetrieveStrategy.java:
```diff
@@ -5,16 +5,16 @@
 import org.springframework.data.domain.Pageable;
 
-// 권한별 업체 목록 조회 전략
+// 권한 별 업체 전체 조회 및 검색 전략 (허브 관리자, 허브 업체)
```
👍 👍

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyFacade.java:
```diff
@@ -2,47 +2,37 @@
+
 @Service
 @RequiredArgsConstructor
 public class CompanyFacade {
```
facade + strategy 패턴을 이용한 권한별 요청 처리 구현 👍 👍

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/strategy/HubCompanyRetrieveStrategy.java:
```diff
@@ -11,37 +11,37 @@
     @Override
-    public boolean isMatched(Role role) { // MASTER_ADMIN
-        return role.isHubAdmin(); // false
+    public boolean isMatched(Role role) { // HUB_COMPANY
```
주석은 빼도 좋을 것 같은데요? 👀

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyFacade.java:
```diff
@@ -2,47 +2,37 @@
-    // ..
-    // ..
-    // ..
+    // HubAdminCompanyRetrieveStrategy - 허브 관리자
```
interface의 구현체들은 어떤 것들이 있는지 확인할 수 있으니, 주석은 삭제해도 좋을 것 같아요!

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyFacade.java:
```diff
@@ -35,4 +37,12 @@ public Page<CompanyResponse> getAllCompaniesForHub(
     }
 
+    // 단일 조회
+    public CompanyResponse getCompanyByIdForHub(Long id, Long currentUser, Role role) {
```
목록 조회 말고도 단일 조회에도 적용해 보셨군요 👍 👍

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/strategy/HubAdminRetrieveByIdStrategy.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public CompanyResponse retrieveCompanyById(Long id, Long currentUser, Role role) {
+        Company company = companyRepo.findByIdAndHubIdAndIsDeletedFalse(id, currentUser)
```
다른 곳에 적용하셨던 것처럼 아래처럼 표현해 볼 수 있을 것 같아요!

```java
        return companyRepo.findByIdAndHubIdAndIsDeletedFalse(id, currentUser)
                .map(CompanyResponse::fromEntity)
                .orElseThrow(() -> new CompanyException(CompanyErrorCode.COMPANY_NOT_FOUND));
```

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/CompanyFacade.java:
```diff
@@ -45,4 +49,13 @@ public CompanyResponse getCompanyByIdForHub(Long id, Long currentUser, Role role
     }
+
+    // 수정
+    public CompanyResponse updateCompany(Long id, UpdateCompanyRequest updateCompanyRequest, Long currentUser, Role role) {
```
제대로 적용 해보시는데요? 굿굿  👍 👍

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/strategy/HubAdminDeleteStrategy.java:
```diff
@@ -0,0 +1,30 @@
+
+        company.delete(); // isDeleted = true로 설정
+        companyRepo.save(company);
+    }
```
@Transactional을 명시하고,

트랜잭션 범위 안에서 EntityManager에 의해 Entity가 관리 될 수 있도록 DirtyChecking을 활용해보시는 것도 좋을 것 같아요!

```java
    @Override
    @Transactional
    public void deleteCompany(Long id, Long currentUser, Role role) {
        Company company = companyRepo.findByIdAndHubIdAndIsDeletedFalse(id, currentUser)
                .orElseThrow(() -> new CompanyException(CompanyErrorCode.UNAUTHORIZED_ACCESS));

        company.delete(); // isDeleted = true로 설정
    }
```

re-mora commented on service/company-app/src/main/java/io/fulflix/company/application/strategy/MasterAdminDeleteStrategy.java:
```diff
@@ -0,0 +1,30 @@
+
+        company.delete(); // isDeleted = true로 설정
+        companyRepo.save(company);
+    }
```
이 부분도 트랜잭션 범위 안에서 Dirty checking을 통해 데이터가 갱신 될 수 있도록 변경해보시면 어떨까요?

re-mora reviewed (commented):
작업량이 많으셨을 텐데 Facade + Strategy 패턴을 잘 적용해주셨네요! 👍 👍 
전반적으로 코드가 많이 깔끔해졌어요 !!

이렇게 비지니스 로직을 분리하거나 개선하는 작업을 진행하면서, 리펙토링이 잘 되고 있는지 확인하는 테스트가 함께 구현되면 작업에 단위가 크더라도, 안정적으로 수정해 볼 수 있을 것 같아요!

다음에는 테스트 케이스 작성에도 도전해 보시면 좋을 것 같습니다!

추가적으로 더 궁금한 내용은 없어서 approve 하도록 하겠습니다!

고생 많으셨어요! 👍 👍

yezyaa committed:
refactor: Optional map 활용하여 코드 간결화

yezyaa committed:
fix: 마스터 관리자 권한으로 업체 삭제 시 리턴 값을 null로 처리한 부분을 isMasterAdmin()으로 수정

yezyaa committed:
refactor: JPA DirtyChecking을 활용하여 Entity가 EntityManager에 의해 관리되도록 수정

yezyaa committed:
Merge remote-tracking branch 'origin/feature/connect-feignclient-to-company' into feature/connect-feignclient-to-company
